### PR TITLE
GH-125603: Don't count executing generators and coroutines as referrers in gc.gc_referrers.

### PR DESCRIPTION
**Pull request by markshannon**
Skipping news as this is reverting behavior that never made it into a release.

* Issue: gh-125603

### CODE REVIEW

**Review by kumaraditya303 — APPROVED** [review]
asyncio changes LGTM

File: Objects/genobject.c [hunk 1]
```diff
@@ -58,7 +58,7 @@ gen_traverse(PyObject *self, visitproc visit, void *arg)
     PyGenObject *gen = _PyGen_CAST(self);
     Py_VISIT(gen->gi_name);
     Py_VISIT(gen->gi_qualname);
-    if (gen->gi_frame_state != FRAME_CLEARED) {
+    if (gen->gi_frame_state < FRAME_EXECUTING) {
```
Comment (colesbury) [inline Objects/genobject.c]: I *think* this is okay, but it also adds some additional constraints that will make the GC in the free threading build more fragile. I think we have to make sure that any gen/coro marked as `FRAME_EXECUTING` is on a `PyThreadState`'s frame stack -- otherwise some deferred references may not be visible to the GC and collected while still in use.

* There must not be any escaping calls between setting `gi_frame_state = FRAME_EXECUTING` and pushing the frame to the thread's stack.
* There must not be any escaping calls between popping the frame from the stack and setting `gi_frame_state` to some other value. (i.e., `exit_unwind` -> `clear_gen_frame`)

Comment (markshannon) [reply Objects/genobject.c]: All those things are already true, because we rely on `gi_frame_state == FRAME_EXECUTING` to guard against sending to an already executing generator.

Comment (colesbury) [reply Objects/genobject.c]: At shutdown we delete all `PyThreadState`s except for the main thread. Those threads could be running generators or coroutines stuck in some long running call (like a `time.sleep()`). We run the GC after that. That seems like it would be unsafe (because we're now hiding deferred `_PyStackRef` from the GC.

Comment (colesbury) [reply Objects/genobject.c]: This test program segfaults in the free threading build with this PR:

https://gist.github.com/colesbury/11d59b9987e881a3c016b086bb4ba1ff

Comment (markshannon) [reply Objects/genobject.c]: If the generator is executing, it is part of the stack, not the heap.
So, I think `tp_traverse` should not be traversing executing generators.

@colesbury 
How does the free-threading GC find references that are on the stack (in normal frames)?
We should do that for executing generators.

Comment (colesbury) [reply Objects/genobject.c]: We loop over every `PyThreadState` in the interpreter and over the stack of frames in the thread state: https://github.com/python/cpython/blob/0cd21406bf84b3b4927a8117024232774823aee0/Python/gc_free_threading.c#L372-L373

During shutdown, we delete all `PyThreadState`s except for the main thread. So even though the generator is still marked as executing, it's not on any thread state's stack (because the `PyThreadState` [got deleted](https://github.com/python/cpython/blob/0cd21406bf84b3b4927a8117024232774823aee0/Python/pylifecycle.c#L2044-L2046) during `_Py_Finalize`)

Comment (markshannon) [reply Objects/genobject.c]: A generator that is marked as executing, but is not any thread's stack, is in an invalid state.

That sounds like the problem to me.
Do we know how it gets in that state?

Comment (colesbury) [reply Objects/genobject.c]: Yes, during shutdown, we delete all `PyThreadState`s except for the main thread.

Comment (markshannon) [reply Objects/genobject.c]: Let me rephrase:
When deleting the thread state, how do we clean up the references?

Presumably, we should be changing the state of the generator when we clean up its frame but we aren't.

Comment (markshannon) [reply Objects/genobject.c]: To answer my own question. We don't cleanup the references 🙁

Comment (colesbury) [reply Objects/genobject.c]: We call `PyThreadState_Clear()` on the deleted thread states, but that [doesn't clean up](https://github.com/python/cpython/blob/3d1df3d84e5c75a52b6f1379cd7f2809fc50befa/Python/pystate.c#L1639-L1651) `current_frame`. We don't call `PyThreadState_Delete()` for reasons that are not clear to me. Even if `PyThreadState_Clear()` cleaned up `current_frame` that wouldn't be sufficient because we unlink all the thread states before calling `PyThreadState_Clear()` and (by your definition) that already puts them in an invalid state. 

This is all longstanding CPython behavior, as far as I can tell. Changing the shutdown behavior seems likely to cause different shutdown-related bugs than what we experience today.

I think we can work around this by encoding more knowledge about generators in `gc_free_threading.c`. I don't love that, but it seems less risky than messing with the shutdown behavior. Let me know if you want to go that route -- I can help with the `gc_free_threading.c` changes.

Comment (markshannon) [reply Objects/genobject.c]: I'd like to investigate fixing up thread state clean up. While it may well risk introducing bugs in the short term, I think to would be better long term. It is hard to optimize anything if we can't trust our supposed invariants.

A quick fix, until we decide on how to handle this long term, would be to traverse the frame list when deleting threads and mark all generators as suspended.

File: Python/pystate.c [hunk 2]
```diff
@@ -1636,18 +1636,28 @@ PyThreadState_Clear(PyThreadState *tstate)
...
+            fprintf(stderr,
+            "PyThreadState_Clear: warning: thread still has a frame\n");
+        }
+        do {
+            if (frame->owner == FRAME_OWNED_BY_GENERATOR) {
+                PyGenObject *gen = _PyGen_GetGeneratorFromFrame(frame);
+                gen->gi_frame_state = FRAME_ZOMBIE;
+            }
+            frame = frame->previous;
+        } while (frame != NULL);
```
Comment (colesbury) [inline Python/pystate.c:1649]: It's not sufficient to mark generators as zombies in `PyThreadState_Clear()`. We call `PyThreadState_Clear()` one at a time on each non-main thread:

* At that point, they are already unlinked from the interpreter's list of `PyThreadStates`
* The `PyThreadState_Clear()` calls makes escaping calls (`Py_CLEAR()`) that can trigger the GC

So the GC may still see generators that are marked as "running", but aren't in any accessible `PyThreadState`'s frame stack.

Comment (colesbury) [reply Python/pystate.c:1649]: I wrote "At that point, they are already unlinked from the interpreter's list of PyThreadStates", but I don't think that's accurate. Thread states are unlinked in `PyThreadState_Delete` or `PyThreadState_DeleteCurrent`, and those happen *after* `PyThreadState_Clear`.

Comment (colesbury) [reply Python/pystate.c:1649]: Oh... my Oct 22, 2024 comment was in relation to interpreter shutdown, which reverses the typical order. During shutdown thread states are:

1) Unlinked
2) Cleared
3) Deleted

Whereas during normal thread exit the order is:

1) Cleared
2) Unlinked
3) Deleted

Specifically, in `_Py_Finalize()` we:

1) First we unlink all the remaining thread states (except the main thread). This is typically for daemon threads, but if you press Ctrl-C it can include non-daemon threads as well:

https://github.com/python/cpython/blob/11503211c6e8985a305f8e2a8530e2258f982180/Python/pylifecycle.c#L2079

2) Then we call `_PyThreadState_DeleteList()`, which calls `PyThreadState_Clear()` on each thread one at a time, which can execute arbitrary code via finalizers and weakrefs.

https://github.com/python/cpython/blob/11503211c6e8985a305f8e2a8530e2258f982180/Python/pylifecycle.c#L2090

https://github.com/python/cpython/blob/11503211c6e8985a305f8e2a8530e2258f982180/Python/pystate.c#L1904-L1922

I think things should be okay if we mark the generators as zombies from `_PyThreadState_RemoveExcept()` as well as from `PyThreadState_Clear()`.

**Comment by markshannon** [comment]
Unlinking a thread before clearing seems strange. Why do we do that during interpreter shutdown?

**Comment by colesbury** [comment]
It looks like we've been doing it this way since 3.9:

1) Looping over the interpreter's linked list of thread states requires `HEAD_LOCK()`
2) It's not safe to call `PyThreadState_Clear()` inside of `HEAD_LOCK()`

Prior to 3.9 we called `PyThreadState_Clear()` inside of `HEAD_LOCK()`, which has it's own problems.

**Comment by sergey-miryanov** [comment]
Is there any way I can help with this?
